trl/trainer/dpo_trainer.py

```diff
@@ -1424,7 +1424,11 @@ def log(self, logs: dict[str, float], start_time: Optional[float] = None) -> Non
         train_eval = "train" if "loss" in logs else "eval"
         # Add averaged stored metrics to logs
         for key, metrics in self._stored_metrics[train_eval].items():
-            logs[key] = torch.tensor(metrics).mean().item()
+            if isinstance(metrics[0], torch.Tensor):
+                gathered = self._nested_gather([m.cuda() for m in metrics])
```

Comment:
do you need `.cuda()`?

Reply:
> do you need `.cuda()`?

metrics are moved cpu before, some backends (e.g. nccl) does not support gathering tensors on cpu. but I admit `.cuda` here loses some genrality. maybe `.to(self.accelerator.device)` is better?

Reply:
> maybe `self.accelerator.gather(metrics).mean().item()` would be simpler?

I agree `self.accelerator.gather` is better. but `metrics` in the loop is a `list[torch.Tensor]` or `list[float]`, so `gather` actually returns a `list[torch.Tensor]`. So I think I change it into:

```python
            if isinstance(metrics[0], torch.Tensor):
                gathered = self.accelerator.gather([m.to(self.accelerator.device) for m in metrics])
                metrics = [g.mean() for g in gathered]
            meaned = torch.tensor(metrics).mean()
            logs[key] = meaned.item()
```

I know creating a new tensor on metrics seems a little weird, but that is how it was originally written. I don't why but I don't why to break anything so I left it there.